Finished SplitDataFrame prep, MLP, Trainer, and Multi-class analysis

ahmedmohammed107 opened the pull request:
The workflow notebook is under the notebooks directory. We can delete it later

sgoldenCS commented on jlab_datascience_toolkit/cfgs/defaults/example_cfg.yaml:
This looks like JSON... While YAML parsers can read JSON files, I feel like we should format our example configurations using real YAML. To do this, I'd suggest simply reading in the file (`cfg = yaml.safe_load(...)`) and then dumping it with PyYAML (`yaml.safe_dump(cfg, filestream)`) which should save the file with YAML styling.

ahmedmohammed107 replied:
Done

sgoldenCS commented on jlab_datascience_toolkit/core/jdst_model.py:
If we are moving these functions to a trainer, should we create a core class for it?

dlersch replied:
yes, the trainer should have a core class.

sgoldenCS commented:
I've looked through the majority of the code and think it looks good (other than my two specific comments about the YAML and trainer core). My only general comments are:
- We should fix the registry import mechanisms so that 
  - all modules are registered whenever the core is imported
  - the registry `make` function is available directly from the core (i.e. `jlab_datascience_toolkit.make()`)
- We should decide on a logger naming convention. For instance, all loggers could be named `"jdst_toolkit.<modulename>"`. This might make it easier when using tools like Hydra, which allow you to enable or disable loggers based on their name. 

Neither of these comments needs to be addressed in this PR, so I would be happy to approve once the earlier comments about the YAML file and trainer core are addressed.

dlersch commented:
I went through your code and I have a few questions:
1. I see that there is a trainer class defined and used in your workflow. Thats good, but why did you then define the fit() function again in the model class ? Is this because of keras ?
2. It would be good to have your workflow as python script in a workflow folder. This way, we have all components at hand

ahmedmohammed107 commented:
> I went through your code and I have a few questions:
> 
> 1. I see that there is a trainer class defined and used in your workflow. Thats good, but why did you then define the fit() function again in the model class ? Is this because of keras ?
> 2. It would be good to have your workflow as python script in a workflow folder. This way, we have all components at hand

1) I agree and I removed the model's fit method. Please note that the KerasMLP class is not a keras.Model itself. Instead, it has a self.model which is keras.Model. When I tried to make KerasMLP inherit from both JDSTModel and keras.Model, I was getting all sorts of error when trying to define self inputs and outputs. In another project, I was doing the following:

class Encoder(keras.Model):
    def __init__(self, configs):
        super().__init__()
        layers_list = get_layers_list(configs)
        self.seq = keras.Sequential(layers_list)
        # Define self inputs & outputs
        inputs = keras.layers.Input(shape=(configs["T"], configs["N"]))
        outputs = self.call(inputs)
        super().__init__(inputs=inputs, outputs=outputs)

    def call(self, x, training=None):
        # x: (B, T_in, N_in)
        return self.seq.call(x, training=training)  # (B, z_size)

But this didn't work when I added JDSTModel to the inherited classes. So I made this workaround which is to define a "model" attribute within the kerasMLP. This is why the trainer's fit method has model.model.compile() and model.model.fit()

2) Done

dlersch commented:
@ahmedmohammed107 , I downloaded your branch and tested all the unit-tests. They are working. Thank you for doing this. Three things that I would like you to fix and then the pull request will be accepted right away:
1. Your workflow uses an explicit path to a config --> Yes you can overwrite the config with argparse, but it would be better to use a relative path to your config, so that any user can run the workflow right away, without having to use the --cfg_fike_path argument
2. I tested your workflow and it works like a charm. Please add one, or two plots to your analysis module. O know, it is just an example, but one or two plots help us making a point of the core / your work. 
3. Please add the seaborn requirement to the .yaml files that are needed to setup the environment. This way everyone has it available 

Thank you so much!

ahmedmohammed107 commented:
> @ahmedmohammed107 , I downloaded your branch and tested all the unit-tests. They are working. Thank you for doing this. Three things that I would like you to fix and then the pull request will be accepted right away:
> 
> 1. Your workflow uses an explicit path to a config --> Yes you can overwrite the config with argparse, but it would be better to use a relative path to your config, so that any user can run the workflow right away, without having to use the --cfg_fike_path argument
> 2. I tested your workflow and it works like a charm. Please add one, or two plots to your analysis module. I know, it is just an example, but one or two plots help us making a point of the core / your work.
> 3. Please add the seaborn requirement to the .yaml files that are needed to setup the environment. This way everyone has it available
> 
> Thank you so much!

@dlersch Done :)
